File: server/camcops_server/tasks/basdai.py
#!/usr/bin/env python

"""
camcops_server/tasks/basdai.py

===============================================================================

    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.

===============================================================================

**Bath Ankylosing Spondylitis Disease Activity Index (BASDAI) task.**

"""

import statistics
from typing import Any, Dict, List, Optional, Type, Tuple

import cardinal_pythonlib.rnc_web as ws
from cardinal_pythonlib.stringfunc import strseq
from sqlalchemy.ext.declarative import DeclarativeMeta
from sqlalchemy.sql.sqltypes import Float

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_db import add_multiple_columns
from camcops_server.cc_modules.cc_html import tr_qa, tr, answer
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_summaryelement import SummaryElement
from camcops_server.cc_modules.cc_task import TaskHasPatientMixin, Task
from camcops_server.cc_modules.cc_trackerhelpers import (
    TrackerAxisTick,
    TrackerInfo,
    TrackerLabel,
)


# =============================================================================
# BASDAI
# =============================================================================


class BasdaiMetaclass(DeclarativeMeta):
    # noinspection PyInitNewSignature
    def __init__(
        cls: Type["Basdai"],
        name: str,
        bases: Tuple[Type, ...],
        classdict: Dict[str, Any],
    ) -> None:

        add_multiple_columns(
            cls,
            "q",
            1,
            cls.N_QUESTIONS,
            coltype=Float,
            minimum=0,
            maximum=10,
            comment_fmt="Q{n} - {s}",
            comment_strings=[
                "fatigue/tiredness 0-10 (none - very severe)",
                "AS neck, back, hip pain 0-10 (none - very severe)",
                "other joint pain/swelling 0-10 (none - very severe)",
                "discomfort from tender areas 0-10 (none - very severe)",
                "morning stiffness level 0-10 (none - very severe)",
                "morning stiffness duration 0-10 (none - 2 or more hours)",
            ],
        )

        super().__init__(name, bases, classdict)


class Basdai(TaskHasPatientMixin, Task, metaclass=BasdaiMetaclass):
    __tablename__ = "basdai"
    shortname = "BASDAI"
    provides_trackers = True

    N_QUESTIONS = 6
    FIELD_NAMES = strseq("q", 1, N_QUESTIONS)

    MINIMUM = 0.0
    ACTIVE_CUTOFF = 4.0
    MAXIMUM = 10.0

    @staticmethod
    def longname(req: "CamcopsRequest") -> str:
        _ = req.gettext
        return _("Bath Ankylosing Spondylitis Disease Activity Index")

    def get_summaries(self, req: CamcopsRequest) -> List[SummaryElement]:
        return self.standard_task_summary_fields() + [
            SummaryElement(
                name="basdai",
                coltype=Float(),
                value=self.basdai(),
                comment="BASDAI",
            )
        ]

    def is_complete(self) -> bool:
        if self.any_fields_none(self.FIELD_NAMES):
            return False

        if not self.field_contents_valid():
            return False

        return True

    def get_trackers(self, req: CamcopsRequest) -> List[TrackerInfo]:
        axis_min = self.MINIMUM - 0.5
        axis_max = self.MAXIMUM + 0.5
        axis_ticks = [
            TrackerAxisTick(n, str(n)) for n in range(0, int(axis_max) + 1)
        ]

        horizontal_lines = [self.MAXIMUM, self.ACTIVE_CUTOFF, self.MINIMUM]

        horizontal_labels = [
            TrackerLabel(
                self.ACTIVE_CUTOFF + 0.5, self.wxstring(req, "active")
            ),
            TrackerLabel(
                self.ACTIVE_CUTOFF - 0.5, self.wxstring(req, "inactive")
            ),
        ]

        return [
            TrackerInfo(
                value=self.basdai(),
                plot_label="BASDAI",
                axis_label="BASDAI",
                axis_min=axis_min,
                axis_max=axis_max,
                axis_ticks=axis_ticks,
                horizontal_lines=horizontal_lines,
                horizontal_labels=horizontal_labels,
            )
        ]

    def basdai(self) -> Optional[float]:
        """
        Calculating the BASDAI
        A. Add scores for questions 1 – 4
        B. Calculate the mean for questions 5 and 6
        C. Add A and B and divide by 5

        The higher the BASDAI score, the more severe the patient’s disability
        due to their AS.
        """
        if not self.is_complete():
            return None

        score_a_field_names = strseq("q", 1, 4)
        score_b_field_names = strseq("q", 5, 6)

        a = sum([getattr(self, q) for q in score_a_field_names])
        b = statistics.mean([getattr(self, q) for q in score_b_field_names])

        return (a + b) / 5

    def activity_state(self, req: CamcopsRequest) -> str:
        basdai = self.basdai()

        if basdai is None:
            return "?"

        if basdai < self.ACTIVE_CUTOFF:
            return self.wxstring(req, "inactive")

        return self.wxstring(req, "active")

    def get_task_html(self, req: CamcopsRequest) -> str:
        rows = ""
        for q_num in range(1, self.N_QUESTIONS + 1):
            q_field = "q" + str(q_num)
            qtext = self.xstring(req, q_field)  # includes HTML
            min_text = self.wxstring(req, q_field + "_min")
            max_text = self.wxstring(req, q_field + "_max")
            qtext += f" <i>(0 = {min_text}, 10 = {max_text})</i>"
            question_cell = f"{q_num}. {qtext}"
            score = getattr(self, q_field)

            rows += tr_qa(question_cell, score)

        basdai = ws.number_to_dp(self.basdai(), 1, default="?")

        html = """
            <div class="{CssClass.SUMMARY}">
                <table class="{CssClass.SUMMARY}">
                    {tr_is_complete}
                    {basdai}
                </table>
            </div>
            <table class="{CssClass.TASKDETAIL}">
                <tr>
                    <th width="60%">Question</th>
                    <th width="40%">Answer</th>
                </tr>
                {rows}
            </table>
            <div class="{CssClass.FOOTNOTES}">
                [1] (A) Add scores for questions 1–4.
                    (B) Calculate the mean for questions 5 and 6.
                    (C) Add A and B and divide by 5, giving a total in the
                        range 0–10.
                    &lt;4.0 suggests inactive disease,
                    &ge;4.0 suggests active disease.
            </div>
        """.format(
            CssClass=CssClass,
            tr_is_complete=self.get_is_complete_tr(req),
            basdai=tr(
                self.wxstring(req, "basdai") + " <sup>[1]</sup>",
                "{} ({})".format(answer(basdai), self.activity_state(req)),
            ),
            rows=rows,
        )
        return html

File: tablet_qt/tasks/basdai.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

#include "basdai.h"

#include "common/uiconst.h"
#include "lib/convert.h"
#include "lib/stringfunc.h"
#include "maths/mathfunc.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/qugridcell.h"
#include "questionnairelib/qugridcontainer.h"
#include "questionnairelib/quslider.h"
#include "questionnairelib/quspacer.h"
#include "questionnairelib/qutext.h"
#include "tasklib/taskfactory.h"
#include "tasklib/taskregistrar.h"
using mathfunc::anyNull;
using stringfunc::strnum;
using stringfunc::strseq;

const int FIRST_Q = 1;
const int N_SCORE_A_QUESTIONS = 4;
const int N_QUESTIONS = 6;
const QString QPREFIX("q");

const int DP = 1;

const QString Basdai::BASDAI_TABLENAME("basdai");

void initializeBasdai(TaskFactory& factory)
{
    static TaskRegistrar<Basdai> registered(factory);
}

Basdai::Basdai(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, BASDAI_TABLENAME, false, false, false),
    // ... anon, clin, resp
    m_questionnaire(nullptr)
{
    addFields(
        strseq(QPREFIX, FIRST_Q, N_QUESTIONS), QMetaType::fromType<double>()
    );

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}

// ============================================================================
// Class info
// ============================================================================

QString Basdai::shortname() const
{
    return "BASDAI";
}

QString Basdai::longname() const
{
    return tr("Bath Ankylosing Spondylitis Disease Activity Index");
}

QString Basdai::description() const
{
    return tr(
        "A self-administered instrument for assessing disease activity in "
        "ankylosing spondylitis"
    );
}

QStringList Basdai::fieldNames() const
{
    return strseq(QPREFIX, FIRST_Q, N_QUESTIONS);
}

QStringList Basdai::scoreAFieldNames() const
{
    return strseq(QPREFIX, FIRST_Q, N_SCORE_A_QUESTIONS);
}

// ============================================================================
// Instance info
// ============================================================================

bool Basdai::isComplete() const
{
    if (anyNull(values(fieldNames()))) {
        return false;
    }

    return true;
}

QVariant Basdai::basdai() const
{
    // Calculating the BASDAI:
    // A. Add scores for questions 1–4
    // B. Calculate the mean for questions 5 and 6
    // C. Add A and B (which is then out of 50) and divide by 5

    // The higher the BASDAI score, the more severe the patient’s disability
    // due to their AS.

    if (!isComplete()) {
        return QVariant();
    }
    const double a = mathfunc::sumDouble(values(scoreAFieldNames()));
    const double b
        = mathfunc::mean(value("q5").toDouble(), value("q6").toDouble());
    return (a + b) / 5;
}

QStringList Basdai::summary() const
{
    return QStringList{QString("%1: %2").arg(
        xstring("basdai"), convert::prettyValue(basdai(), DP)
    )};
}

QStringList Basdai::detail() const
{
    QStringList lines = completenessInfo();
    const QString spacer = " ";
    const QString suffix = "";
    lines
        += fieldSummaries("q", suffix, spacer, QPREFIX, FIRST_Q, N_QUESTIONS);
    lines.append("");
    lines += summary();

    return lines;
}

OpenableWidget* Basdai::editor(const bool read_only)
{
    QuPagePtr page((new QuPage{new QuText(xstring("instructions"))})
                       ->setTitle(xstring("title_main")));

    auto slider_grid = new QuGridContainer();
    slider_grid->setExpandHorizontally(false);
    slider_grid->setFixedGrid(false);

    page->addElement(slider_grid);

    const int QUESTION_ROW_SPAN = 1;
    const int QUESTION_COLUMN_SPAN = 3;
    const Qt::Alignment align_right = Qt::AlignRight | Qt::AlignVCenter;
    const Qt::Alignment align_left = Qt::AlignLeft | Qt::AlignVCenter;
    const Qt::Alignment align_centre = Qt::AlignCenter | Qt::AlignVCenter;
    const int row_span = 1;
    const int col_span = 1;
    const double int_max = 100;
    const double big_step = int_max / 10;
    const double float_max = 10.0;
    const int dp = 1;
    const QMap<int, QString> final_q_extra_labels{
        {int_max * 0.25, "½"},
        {int_max * 0.5, "1"},
        {int_max * 0.75, "1½"},
    };
    const bool can_shrink = true;

    int row = 0;
    const auto fieldnames = fieldNames();
    for (const QString& fieldname : fieldnames) {
        const bool last_q = fieldname == strnum(QPREFIX, 6);

        // Represent a float in the range 0-10 to 1dp, via integers in the
        // range 0-100.
        QuSlider* slider = new QuSlider(fieldRef(fieldname), 0, int_max, 1);
        slider->setConvertForRealField(true, 0.0, float_max, dp);
        slider->setHorizontal(true);
        slider->setBigStep(big_step);
        slider->setAbsoluteLengthCm(10, can_shrink);
        if (last_q) {
            slider->setTickLabels(final_q_extra_labels);
            slider->setTickLabelPosition(QSlider::TicksBelow);
        } else {
            slider->setTickLabelPosition(QSlider::NoTicks);
        }
        slider->setShowValue(false);
        slider->setSymmetric(true);

        const auto question_text = new QuText(xstring(fieldname));
        slider_grid->addCell(QuGridCell(
            question_text, row, 0, QUESTION_ROW_SPAN, QUESTION_COLUMN_SPAN
        ));
        row++;

        const auto min_label = new QuText(xstring(fieldname + "_min"));
        min_label->setTextAndWidgetAlignment(align_right);
        const auto max_label = new QuText(xstring(fieldname + "_max"));
        max_label->setTextAndWidgetAlignment(align_left);
        slider_grid->addCell(
            QuGridCell(min_label, row, 0, row_span, col_span, align_right)
        );
        slider_grid->addCell(
            QuGridCell(slider, row, 1, row_span, col_span, align_centre)
        );
        slider_grid->addCell(
            QuGridCell(max_label, row, 2, row_span, col_span, align_left)
        );
        row++;

        if (!last_q) {
            slider_grid->addCell(QuGridCell(
                new QuSpacer(QSize(uiconst::BIGSPACE, uiconst::BIGSPACE)),
                row,
                0
            ));
            row++;
        }
    }

    m_questionnaire = new Questionnaire(m_app, {page});
    m_questionnaire->setType(QuPage::PageType::Patient);
    m_questionnaire->setReadOnly(read_only);

    return m_questionnaire;
}
